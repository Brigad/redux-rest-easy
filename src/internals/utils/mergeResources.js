import isObject from 'lodash.isplainobject';
import mergeWith from 'lodash.mergewith';

const mergeResources = (state, normalizedPayload) =>
  mergeWith(
    state.resources || {},
    normalizedPayload,
    (resources, newResources) =>
      isObject(resources) ? { ...resources, ...newResources } : undefined,
  );

export default mergeResources;
