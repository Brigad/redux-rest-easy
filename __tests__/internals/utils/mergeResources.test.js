import mergeResources from '../../../src/internals/utils/mergeResources';

const EMPTY_STATE = {};
const FILLED_STATE = {
  resources: {
    fruits: {
      apple: {
        juicy: true,
        red: true,
      },
      cherry: {
        juicy: false,
      },
    },
    vegetables: {
      carrot: {
        orange: true,
      },
    },
  },
};
const FILLED_STATE2 = {
  resources: {
    fruits: {
      apple: {
        juicy: false,
      },
      banana: {
        yellow: true,
      },
    },
  },
};

describe('mergeResources', () => {
  test('merge filled in empty', () => {
    expect(
      mergeResources(EMPTY_STATE, FILLED_STATE.resources),
    ).toMatchSnapshot();
  });

  test('merge empty in filled', () => {
    expect(
      mergeResources(FILLED_STATE, EMPTY_STATE.resources),
    ).toMatchSnapshot();
  });

  test('merge filled2 in filled', () => {
    expect(
      mergeResources(FILLED_STATE, FILLED_STATE2.resources),
    ).toMatchSnapshot();
  });
});
