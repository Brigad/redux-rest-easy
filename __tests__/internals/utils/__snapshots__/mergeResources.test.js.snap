// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mergeResources merge empty in filled 1`] = `
Object {
  "fruits": Object {
    "apple": Object {
      "juicy": true,
      "red": true,
    },
    "cherry": Object {
      "juicy": false,
    },
  },
  "vegetables": Object {
    "carrot": Object {
      "orange": true,
    },
  },
}
`;

exports[`mergeResources merge filled in empty 1`] = `
Object {
  "fruits": Object {
    "apple": Object {
      "juicy": true,
      "red": true,
    },
    "cherry": Object {
      "juicy": false,
    },
  },
  "vegetables": Object {
    "carrot": Object {
      "orange": true,
    },
  },
}
`;

exports[`mergeResources merge filled2 in filled 1`] = `
Object {
  "fruits": Object {
    "apple": Object {
      "juicy": false,
    },
    "banana": Object {
      "yellow": true,
    },
    "cherry": Object {
      "juicy": false,
    },
  },
  "vegetables": Object {
    "carrot": Object {
      "orange": true,
    },
  },
}
`;
